### PR TITLE
Updates POD section in workload get output same as Kubectl get

### LINKED ISSUE #224: Workload get output pod status should mimic kubectl output
### Description of problem

`tanzu apps workload get` shows the status of `Pending` when a build is running. The status `Pending` while accurate, is not very helpful. 

Screen Shot 2022-07-17 at 11 02 38 PM

### Proposed solution
`apps` plugin should mimic the output of `kubectl` as shown in the screenshot below. It should show `init` container information when the pods are not in a `completed` or `succeeded` status.

Screen Shot 2022-07-17 at 11 04 55 PM

cc @danfein for design recommendations.

### PR DESCRIPTION
**Pull request by shaheerkootteeri**
Pull request

### What this PR does / why we need it
PR will display the POD details same as that of kubectl get pod output command results.

### Which issue(s) this PR fixes

Fixes #224 

### Describe testing done for PR

Unit test cases added, also updated existing test cases to reflect the POD details output. 
Tested in local cluster.

### CODE REVIEW

**Comment by shaheerkootteeri** [comment]
image

**Comment by shaheerkootteeri** [comment]
image

File: pkg/cli-runtime/client.go [hunk 1]
```diff
@@ -66,6 +72,19 @@ func (c *client) Discovery() discovery.DiscoveryInterface {
...
 }
+func (c *client) ToRESTConfig() (*rest.Config, error) {
+	return c.KubeRestConfig(), nil
+}
+
+func (c *client) ToDiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
+	return disk.NewCachedDiscoveryClientForConfig(c.restConfig, "", "", 0) // need alignment with sash
+}
+func (c *client) ToRawKubeConfigLoader() clientcmd.ClientConfig {
+	return nil
```
Comment (odinnordico) [inline pkg/cli-runtime/client.go:83]: Is it fine to return `nil` here, is `ToRawKubeConfigLoader()`going to be used? if not why is it added to the interface?

Comment (shaheerkootteeri) [reply pkg/cli-runtime/client.go:83]: We are not using it and I have updated the code to remove it from interface as well as implementation

File: pkg/cli-runtime/options.go [hunk 2]
```diff
@@ -29,10 +29,12 @@ import (
 // validated. This function is typically used to define the PreRunE phase of a command.
 //
 // ```
-// cmd := &cobra.Command{
-// 	   ...
-// 	   PreRunE: cli.ValidateE(obj),
-// }
+//
+//	cmd := &cobra.Command{
```
Comment (shashwathi) [inline pkg/cli-runtime/options.go:33]: Remove changes in this file. It is not related to the PR

File: pkg/cli-runtime/printer/table/tablegenerator.go [hunk 3]
```diff
@@ -154,7 +154,9 @@ func (h *HumanReadablePrinter) TableHandler(columnDefinitions []metav1beta1.Tabl
 // ValidateRowPrintHandlerFunc validates print handler signature.
 // printFunc is the function that will be called to print an object.
 // It must be of the following type:
-//  func printFunc(object ObjectType, options PrintOptions) ([]metav1beta1.TableRow, error)
+//
+//	func printFunc(object ObjectType, options PrintOptions) ([]metav1beta1.TableRow, error)
```
Comment (shashwathi) [inline pkg/cli-runtime/printer/table/tablegenerator.go:158]: Remove changes in this file. It is not related to the PR

File: pkg/cli-runtime/printer/tabwriter/tabwriter.go [hunk 4]
```diff
@@ -95,7 +94,6 @@ type cell struct {
 // The Writer must buffer input internally, because proper spacing
 // of one line may depend on the cells in future lines. Clients must
 // call Flush when done calling Write.
```
Comment (shashwathi) [inline pkg/cli-runtime/printer/tabwriter/tabwriter.go:96]: remove all newly added new lines in this PR

File: pkg/cli-runtime/testing/client.go [hunk 5]
```diff
@@ -50,7 +59,132 @@ func NewFakeCliClient(c crclient.Client) cli.Client {
...
+	// Swap out the HTTP client out of the client with the fake's version.
+	fakeClient := c.Clients.(*fake.RESTClient)
+	restClient, err := rest.RESTClientFor(c.ClientConfigVal)
+	if err != nil {
+		panic(err)
+	}
+	restClient.Client = fakeClient.Client
+	return restClient, nil
+}
+func (c *fakeclient) RESTMapper() meta.RESTMapper {
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/client.go:73]: add a new line between each func

File: pkg/cli-runtime/testing/client.go [hunk 6]
```diff
@@ -50,7 +59,132 @@ func NewFakeCliClient(c crclient.Client) cli.Client {
...
+	return testRESTMapper()
+}
+func (c *fakeclient) ToRESTMapper() (meta.RESTMapper, error) {
+	return c.RESTMapper(), nil
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/client.go:77]: why are there 2 func `ToRESTMapper` and `RESTMapper`?

File: pkg/cli-runtime/testing/client.go [hunk 7]
```diff
@@ -50,7 +59,132 @@ func NewFakeCliClient(c crclient.Client) cli.Client {
...
+				PreferredVersion: metav1.GroupVersionForDiscovery{Version: "v1"},
+			},
+			VersionedResources: map[string][]metav1.APIResource{
+				"v1": {
+					{Name: "pods", Namespaced: true, Kind: "Pod"},
+					{Name: "services", Namespaced: true, Kind: "Service"},
+					{Name: "replicationcontrollers", Namespaced: true, Kind: "ReplicationController"},
+					{Name: "componentstatuses", Namespaced: false, Kind: "ComponentStatus"},
+					{Name: "nodes", Namespaced: false, Kind: "Node"},
+					{Name: "secrets", Namespaced: true, Kind: "Secret"},
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/client.go:166]: nit: start with set of API resources we are using for tests.

File: pkg/cli-runtime/testing/command_table.go [hunk 8]
```diff
@@ -223,7 +234,30 @@ func (tc CommandTestCase) Run(t *testing.T, scheme *runtime.Scheme, cmdFactory f
 				t.Errorf("error during prepare: %s", err)
 			}
 		}
+		if len(tc.GivenObjects) != 0 {
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/command_table.go:237]: remove the if condition
```
var pods []client.Object
for loop {
}
if len(pods) > 0 {
   fake builder
}
```

File: pkg/cli-runtime/testing/client.go [hunk 9]
```diff
@@ -50,7 +59,132 @@ func NewFakeCliClient(c crclient.Client) cli.Client {
...
+		Invalidations:      0,
+	}
+}
+func (d *FakeCachedDiscoveryClient) ServerGroupsAndResources() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	return d.Groups, d.Resources, nil
+}
+
+// NewBuilder returns a new resource builder for structured api objects.
+func (c *fakeclient) NewBuilder() *resource.Builder {
+	return resource.NewFakeBuilder(
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/client.go:99]: no test is accessing fake builder from this func so why do you need this? return `panic(not implemented)` for this func

File: pkg/commands/workload_get.go [hunk 10]
```diff
@@ -218,23 +219,40 @@ func (opts *WorkloadGetOptions) Exec(ctx context.Context, c *cli.Config) error {
...
+	arg := []string{"Pod"}
+	r := c.Builder.Unstructured().
+		NamespaceParam(workload.Namespace).
+		LabelSelectorParam(fmt.Sprintf("%s%s%s", cartov1alpha1.WorkloadLabelName, "=", workload.Name)).
+		ResourceTypeOrNameArgs(true, arg...).
+		Latest().
+		Flatten().
+		TransformRequests(func(req *rest.Request) {
+			req.SetHeader("Accept", strings.Join([]string{
+				fmt.Sprintf("application/json;as=Table;v=%s;g=%s", metav1.SchemeGroupVersion.Version, metav1.GroupName),
```
Comment (shashwathi) [inline pkg/commands/workload_get.go:231]: declare  `application/json` as const

File: pkg/commands/workload_get.go [hunk 11]
```diff
@@ -218,23 +219,40 @@ func (opts *WorkloadGetOptions) Exec(ctx context.Context, c *cli.Config) error {
...
+		}).
+		Do()
+	infos, err := r.Infos()
 	if err != nil {
 		c.Eprintf("\n")
 		c.Eerrorf("Failed to list pods:\n")
 		c.Eprintf("  %s\n", err)
 	} else {
-		if len(pods.Items) == 0 {
+		if len(infos) == 0 {
```
Comment (shashwathi) [inline pkg/commands/workload_get.go:242]: This logic needs to be redone
- read the infos object and decide podList obj to be printed
- if podList lenght is empty then print `No pods...` msg 
else print the list of pods.
- Looks like you are expecting infos to have 1 podList object in this case so explicitly call `				tableResult, podlist, _ := printer.DecodeIntoTable(infos[i0].Object)
`

File: pkg/printer/pod_printer.go [hunk 12]
```diff
@@ -80,3 +92,93 @@ func maxContainerRestarts(status corev1.PodStatus) int {
 	}
 	return maxRestarts
 }
+func PodTablePrinterFromObject(c *cli.Config, podlist []Podview, tableResult runtime.Object) error {
```
Comment (shashwathi) [inline pkg/printer/pod_printer.go:95]: Remove the other pod printer func

**Review by shashwathi — COMMENTED** [review]
this looks good
- add e2e test 
 For this test create a pod with workload label and add it to kind cluster. Expect in the get output that it contains the table column definitions or the pod table output.

File: pkg/printer/pod_printer.go (hunk 12 above)
Comment (shaheerkootteeri) [reply pkg/printer/pod_printer.go]: Done

File: pkg/commands/workload_get.go (hunk 11 above)
Comment (shaheerkootteeri) [reply pkg/commands/workload_get.go]: Done

File: pkg/cli-runtime/testing/command_table.go (hunk 8 above)
Comment (shaheerkootteeri) [reply pkg/cli-runtime/testing/command_table.go]: We would need a default builder as the vendor k8s file is throwing null pointer exception if we do not pass a builder instance when calling unstructured function.

File: pkg/cli-runtime/testing/client.go (hunk 6 above)
Comment (shaheerkootteeri) [reply pkg/cli-runtime/testing/client.go]: Removed Restmapper

File: pkg/cli-runtime/testing/client.go (hunk 5 above)
Comment (shaheerkootteeri) [reply pkg/cli-runtime/testing/client.go]: Done

File: pkg/cli-runtime/printer/tabwriter/tabwriter.go (hunk 4 above)
Comment (shaheerkootteeri) [reply pkg/cli-runtime/printer/tabwriter/tabwriter.go]: Done

File: pkg/cli-runtime/printer/table/tablegenerator.go (hunk 3 above)
Comment (shaheerkootteeri) [reply pkg/cli-runtime/printer/table/tablegenerator.go]: Done

File: pkg/cli-runtime/options.go (hunk 2 above)
Comment (shaheerkootteeri) [reply pkg/cli-runtime/options.go]: Done

**Comment by shaheerkootteeri** [comment]
closing this PR since a new PR 292 has been created for the same with the latest code. All the review comments are also being taken care as part of the new PR.

**Pull request closed without merging**